File: Data Structures & Algorithms/matrixDFS/submission-0.java
class Solution {
    public int countPaths(int[][] grid) {
    
    int r = grid.length;
    int c = grid[0].length;
    int [][] visit = new int [r][c];
    return dfs(grid, 0,0,visit);

   
    }

    public int dfs(int[][]grid, int row, int col, int[][]visit)
    {
        if (row< 0 || col<0 || row == grid.length || col == grid[0].length ||
            visit[row][col] == 1 || grid[row][col] == 1 ) {
            return 0;
        }
        if(row == grid.length-1 && col == grid[0].length-1)
        {
            return 1;
        }

        visit[row][col] = 1;
        int count =0;
        count+= dfs(grid,row+1,col,visit);
        count+= dfs(grid,row-1,col,visit);
        count+= dfs(grid,row,col+1,visit);
        count+= dfs(grid,row,col-1,visit);

        visit[row][col]= 0;
        return count;
    }
}

File: Data Structures & Algorithms/search-for-word/submission-0.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        visit = set()
        row = len(board)
        col = len(board[0])

        def dfs(r,c,i):
            if i == len(word):
                return True
            
            if r<0 or c<0 or r>=row or c>=col or (r,c) in visit or word[i] != board[r][c]:
                return False
            
            visit.add((r,c))
            res = dfs(r+1,c,i+1)or dfs(r-1,c,i+1) or dfs(r,c+1,i+1) or dfs(r,c-1,i+1)
            visit.remove((r,c))
            return res


        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i,j,0):
                    return True
        return False